### DISCUSSION TITLE
hostname not set using cloud init snippet

### DISCUSSION THREAD

**Discussion by Surfjoan** [bpg/terraform-provider-proxmox#1666] [Q&A] [answered]
Hi. I have been testing this provider to spin up vms with opentofu. i have created a proxmox_virtual_environment_file with a custom cloud init file inside and it works to spin up the vms the way i want. but i cannot figure out how the hostname is set? i am using a cloud image from ubuntu and hostname defaults to ubuntu even if i set the vm name in proxmox to something else. I am using one snippet file to spin up all vms so i cannot set something static there. below is the 
proxmox_virtual_environment_file. My understanding is that the hostname should be set to the same name as the vm has in proxmox??

```
resource "proxmox_virtual_environment_file" "ubuntu_cloud_init" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = "proxmox"

  source_raw {
    data = <<EOF
#cloud-config
chpasswd:
  list: |
    ubuntu:ubuntu
  expire: false
package_upgrade: true
packages:
  - qemu-guest-agent
timezone: Europe/Stockholm

users:
  - default
  - name: ubuntu
    groups: sudo
    shell: /bin/bash
    ssh-authorized-keys:
      - ${trimspace("ssh-rsa ff yy@dd")}
    sudo: ALL=(ALL) NOPASSWD:ALL

power_state:
    delay: now
    mode: reboot
    message: Rebooting after cloud-init completion
    condition: true

EOF

    file_name = "ubuntu.cloud-config.yaml"
  }
}
```

**Comment by loganmancuso** [comment, marked as answer]
I think you might just be missing the 'hostname:' parameter in your yaml, here is my default cloud-init snippet. The line `hostname: ${var.instance.name}.local` sets the name. 

```
resource "proxmox_virtual_environment_file" "cloudinit" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = local.node.name

  source_raw {
    file_name = "${var.instance.name}.cloud-config.yaml"
    data      = <<EOT
#cloud-config
hostname: ${var.instance.name}.local
users:
  - name: ${local.credentials_instance.username}
    primary_group: ${local.credentials_instance.username}
    password: "${local.credentials_instance.hashed_password}"
    lock_passwd: false
    ssh-authorized-keys:
      - ${trimspace(local.credentials_instance.sshkey_public)}
    sudo: ALL=(ALL) NOPASSWD:ALL
package_update: false
package_upgrade: false
runcmd:
  - [ sh, -c, "echo $(date) ': Resizing LVM'" ]
  - [ sh, -c, "growpart /dev/vda 3" ]
  - [ sh, -c, "pvresize /dev/vda3" ]
  - [ sh, -c, "lvextend -l +100%FREE $(lvs --noheadings -o lv_name,lv_path | grep $(vgs --noheadings -o vg_name | awk '{print $1}') | awk '{print $2}')" ]
  - [ sh, -c, "resize2fs $(lvs --noheadings -o lv_name,lv_path | grep $(vgs --noheadings -o vg_name | awk '{print $1}') | awk '{print $2}')" ]
EOT
  }
}
```

**Reply by bpg-dev** [reply]
Thanks, @loganmancuso, that’s correct. If a VM is provided with a custom cloud-init, it overrides the default one generated by PVE, which sets the hostname to match the VM name. So, the custom cloud-init has to do to that now.

There’s another example [here](https://github.com/bpg/terraform-provider-proxmox/issues/1094#issuecomment-1979814366).
